File: csharp/EgyNames/DataLoader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Reflection;
using System.Text.Json;

namespace EgyptianNames
{
    public class DataBundle
    {
        public string Version { get; set; } = "0.1.0";
        public int CorpusTokens { get; set; }
        public int CorpusStudents { get; set; }
        public List<int> CohortYears { get; set; } = new List<int>();
        public List<NameEntry> Names { get; set; } = new List<NameEntry>();
        public Dictionary<string, string> Corrections { get; set; } = new Dictionary<string, string>();
    }

    public static class DataLoader
    {
        private static DataBundle? _cachedBundle;
        private static readonly object _lock = new object();

        public static DataBundle LoadBundle(string? customPath = null)
        {
            if (_cachedBundle != null && customPath == null)
            {
                return _cachedBundle;
            }

            lock (_lock)
            {
                if (_cachedBundle != null && customPath == null)
                {
                    return _cachedBundle;
                }

                Stream? stream = null;

                if (!string.IsNullOrEmpty(customPath) && File.Exists(customPath))
                {
                    stream = File.OpenRead(customPath);
                }
                else
                {
                    // Check local relative directories
                    var possiblePaths = new[]
                    {
                        "Data/names.json.gz",
                        "names.json.gz",
                        Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Data", "names.json.gz"),
                        Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "names.json.gz")
                    };

                    foreach (var p in possiblePaths)
                    {
                        if (File.Exists(p))
                        {
                            stream = File.OpenRead(p);
                            break;
                        }
                    }

                    // Fallback to Embedded Resource
                    if (stream == null)
                    {
                        var assembly = Assembly.GetExecutingAssembly();
                        var resourceName = assembly.GetManifestResourceNames()
                            .FirstOrDefault(r => r.EndsWith("names.json.gz", StringComparison.OrdinalIgnoreCase));

                        if (resourceName != null)
                        {
                            stream = assembly.GetManifestResourceStream(resourceName);
                        }
                    }
                }

                if (stream == null)
                {
                    throw new FileNotFoundException("Could not locate names.json.gz data file as embedded resource or file.");
                }

                using (stream)
                using (var gzip = new GZipStream(stream, CompressionMode.Decompress))
                using (var doc = JsonDocument.Parse(gzip))
                {
                    var root = doc.RootElement;
                    var bundle = new DataBundle
                    {
                        Version = root.TryGetProperty("version", out var v) ? v.GetString() ?? "0.1.0" : "0.1.0",
                        CorpusTokens = root.TryGetProperty("corpus_tokens", out var ct) ? ct.GetInt32() : 0,
                        CorpusStudents = root.TryGetProperty("corpus_students", out var cs) ? cs.GetInt32() : 0
                    };

                    if (root.TryGetProperty("cohort_years", out var cy))
                    {
                        foreach (var y in cy.EnumerateArray())
                        {
                            bundle.CohortYears.Add(y.GetInt32());
                        }
                    }

                    if (root.TryGetProperty("corrections", out var corr))
                    {
                        foreach (var prop in corr.EnumerateObject())
                        {
                            bundle.Corrections[prop.Name] = prop.Value.GetString() ?? string.Empty;
                        }
                    }

                    if (root.TryGetProperty("names", out var namesArr))
                    {
                        foreach (var elem in namesArr.EnumerateArray())
                        {
                            var ar = elem.GetProperty("a").GetString() ?? string.Empty;
                            var en = elem.GetProperty("e").GetString() ?? string.Empty;
                            var gStr = elem.TryGetProperty("g", out var g) ? g.GetString() : "n";
                            var rStr = elem.TryGetProperty("r", out var r) ? r.GetString() : "n";
                            var lStr = elem.TryGetProperty("l", out var l) ? l.GetString() : "g";
                            var avStr = elem.TryGetProperty("av", out var av) ? av.GetString() : string.Empty;
                            var evStr = elem.TryGetProperty("ev", out var ev) ? ev.GetString() : string.Empty;
                            var fcStr = elem.TryGetProperty("fc", out var fc) ? fc.GetString() : "r";
                            var tStr = elem.TryGetProperty("t", out var t) ? t.GetString() : string.Empty;
                            var maStr = elem.TryGetProperty("ma", out var ma) ? ma.GetString() : string.Empty;
                            var meStr = elem.TryGetProperty("me", out var me) ? me.GetString() : string.Empty;
                            var tp = elem.TryGetProperty("tp", out var tpVal) ? tpVal.GetDouble() : 0.0;

                            var pList = new List<double>();
                            if (elem.TryGetProperty("p", out var pArr))
                            {
                                foreach (var pv in pArr.EnumerateArray())
                                {
                                    pList.Add(pv.GetDouble());
                                }
                            }

                            var gender = gStr == "m" ? Gender.Male : (gStr == "f" ? Gender.Female : Gender.Neutral);
                            var religion = rStr == "m" ? Religion.Muslim : (rStr == "c" ? Religion.Christian : Religion.Neutral);
                            var role = lStr == "f" ? NameRole.Family : NameRole.Given;
                            var freq = fcStr == "c" ? FrequencyClass.Common : (fcStr == "n" ? FrequencyClass.Normal : FrequencyClass.Rare);

                            var arVariants = !string.IsNullOrEmpty(avStr)
                                ? avStr.Split('|', StringSplitOptions.RemoveEmptyEntries).ToList()
                                : new List<string> { ar };

                            var enVariants = !string.IsNullOrEmpty(evStr)
                                ? evStr.Split('|', StringSplitOptions.RemoveEmptyEntries).ToList()
                                : new List<string> { en };

                            bundle.Names.Add(new NameEntry
                            {
                                Ar = ar,
                                En = en,
                                Gender = gender,
                                Religion = religion,
                                Role = role,
                                ArVariants = arVariants,
                                EnVariants = enVariants,
                                SlotPcts = pList,
                                CorpusShare = tp,
                                Frequency = freq,
                                Tashkeel = tStr ?? string.Empty,
                                MeaningAr = maStr ?? string.Empty,
                                MeaningEn = meStr ?? string.Empty
                            });
                        }
                    }

                    if (customPath == null)
                    {
                        _cachedBundle = bundle;
                    }

                    return bundle;
                }
            }
        }
    }
}

File: java/egy-names/src/main/java/com/afify/egynames/data/DataLoader.java
package com.afify.egynames.data;

import com.afify.egynames.model.Models;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.util.*;
import java.util.zip.GZIPInputStream;

public class DataLoader {

    public static class DataBundle {
        public String version;
        public int corpusTokens;
        public int corpusStudents;
        public List<Integer> cohortYears = new ArrayList<>();
        public List<Models.NameEntry> names = new ArrayList<>();
        public Map<String, String> corrections = new HashMap<>();
    }

    private static DataBundle cachedBundle = null;

    public static synchronized DataBundle loadBundle(String customPath) {
        if (cachedBundle != null && customPath == null) {
            return cachedBundle;
        }

        try {
            InputStream is = null;
            if (customPath != null && new File(customPath).exists()) {
                is = new FileInputStream(customPath);
            } else {
                is = DataLoader.class.getClassLoader().getResourceAsStream("names.json.gz");
                if (is == null) {
                    is = DataLoader.class.getResourceAsStream("/names.json.gz");
                }
                if (is == null && new File("src/main/resources/names.json.gz").exists()) {
                    is = new FileInputStream("src/main/resources/names.json.gz");
                }
            }

            if (is == null) {
                throw new IllegalStateException("Could not find names.json.gz resource or file.");
            }

            try (GZIPInputStream gzip = new GZIPInputStream(is)) {
                ObjectMapper mapper = new ObjectMapper();
                JsonNode root = mapper.readTree(gzip);

                DataBundle bundle = new DataBundle();
                bundle.version = root.has("version") ? root.get("version").asText() : "0.1.0";
                bundle.corpusTokens = root.has("corpus_tokens") ? root.get("corpus_tokens").asInt() : 0;
                bundle.corpusStudents = root.has("corpus_students") ? root.get("corpus_students").asInt() : 0;

                if (root.has("cohort_years")) {
                    for (JsonNode y : root.get("cohort_years")) {
                        bundle.cohortYears.add(y.asInt());
                    }
                }

                if (root.has("corrections")) {
                    Iterator<Map.Entry<String, JsonNode>> fields = root.get("corrections").fields();
                    while (fields.hasNext()) {
                        Map.Entry<String, JsonNode> field = fields.next();
                        bundle.corrections.put(field.getKey(), field.getValue().asText());
                    }
                }

                if (root.has("names")) {
                    for (JsonNode elem : root.get("names")) {
                        Models.NameEntry entry = new Models.NameEntry();
                        entry.ar = elem.has("a") ? elem.get("a").asText() : "";
                        entry.en = elem.has("e") ? elem.get("e").asText() : "";
                        entry.gender = Models.Gender.fromCode(elem.has("g") ? elem.get("g").asText() : "n");
                        entry.religion = Models.Religion.fromCode(elem.has("r") ? elem.get("r").asText() : "n");
                        entry.role = Models.NameRole.fromCode(elem.has("l") ? elem.get("l").asText() : "g");
                        entry.frequency = Models.FrequencyClass.fromCode(elem.has("fc") ? elem.get("fc").asText() : "r");
                        entry.tashkeel = elem.has("t") ? elem.get("t").asText() : "";
                        entry.meaningAr = elem.has("ma") ? elem.get("ma").asText() : "";
                        entry.meaningEn = elem.has("me") ? elem.get("me").asText() : "";
                        entry.corpusShare = elem.has("tp") ? elem.get("tp").asDouble() : 0.0;

                        String av = elem.has("av") ? elem.get("av").asText() : "";
                        String ev = elem.has("ev") ? elem.get("ev").asText() : "";

                        entry.arVariants = !av.isEmpty() ? Arrays.asList(av.split("\\|")) : List.of(entry.ar);
                        entry.enVariants = !ev.isEmpty() ? Arrays.asList(ev.split("\\|")) : List.of(entry.en);

                        List<Double> pList = new ArrayList<>();
                        if (elem.has("p")) {
                            for (JsonNode pv : elem.get("p")) {
                                pList.add(pv.asDouble());
                            }
                        }
                        entry.slotPcts = pList;

                        bundle.names.add(entry);
                    }
                }

                if (customPath == null) {
                    cachedBundle = bundle;
                }
                return bundle;
            }
        } catch (Exception e) {
            throw new RuntimeException("Failed to load EgyptianNames dataset", e);
        }
    }
}
